### PR TITLE
Proposal: Standardizing how new scheme modules are added.

### PR DESCRIPTION
**Pull request by AmeBel**
**Why?** 
1. To avoid having to install modulerized scheme code for every change during developement
2. To avoid having to choose between two ways of defining modules (specifically their installation path)
  * in `/repo-name/openocg/scm/some-module.scm`, or
  * in `/repo-name/opencog/some-module/some-module.scm` .

3 To start using `(use-modules (opencog some-module))` for loading scheme code instead of having [this](https://github.com/AmeBel/opencog/blob/master/lib/development.conf#L114-L142). (Note: this doesn't address https://github.com/opencog/atomspace/issues/508)

**Solution Overivew**
1. Roses standardeized way of defining new packages was used as an example approach.
2. Directory structure for defining new scheme module have two options
  * `/repo-name/opencog/some-module/some-module.scm` imported by `(use-modules (opencog some-module ))`
  * `/repo-name/opencog/some-module/scm/some-module.scm` imported by `(use-modules (opencog some-module))`. 
  * `/repo-name/opencog/some-module/some-module/sub-module.scm` imported by `(use-modules (opencog some-module sub-module))`
  * `/repo-name/opencog/some-module/scm/some-module/sub-module.scm` imported by `(use-modules (opencog some-module sub-module))`. This approach is basically for cases where there are python/c++ modules. If the `sub-module` has a python/c++ code adding`scm` directory under it isn't supported, because we don't many levels deep of submodules.

3 During cmake  configuration stage, the files added as modules are,
  * symlinked in the build directory such that it would have the directory structure to allow import from the cogserver scheme-shell after running `(add-to-load-path "/path/to/build/directory/opencog/scm")` as if the they were installed.
  * configured for installation in the appropriate directory depending on their names

4 The relex2logic module have been used as an example case.

**If accepted then what?**
* Restructure the rest of the scheme modules in this and atomspace repo.
* Do the same for python modules.
* Document it on the wiki and add examples.

### CODE REVIEW

**Comment by AmeBel** [comment]
@linas feedback please.

File: lib/OpenCogFunctions.cmake [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        SET(MODULE_FILE_DIR_PATH ${CMAKE_MATCH_2})
+        SET(MODULE_FILES_DIR_PATH ${CMAKE_MATCH_2}/${CMAKE_MATCH_3})
+        SET(GUILE_SYMLINK_DIR "${CMAKE_BINARY_DIR}/opencog/scm")
+        SET(GUILE_INSTALL_DIR "${DATADIR}/scm")
+
+        # Create symlinks in build directory mirroring the install path structure.
+        # Also configure for install.
+        IF ("${MODULE_NAME}.scm" STREQUAL "${MODULE_FILE}")
+            EXECUTE_PROCESS(
+                COMMAND ${CMAKE_COMMAND} -E make_directory ${CMAKE_BINARY_DIR}/opencog/scm/${MODULE_FILE_DIR_PATH}
```
Comment (linas) [inline lib/OpenCogFunctions.cmake]: you can just say `make_directory ${GUILE_SYMLINK_DIR}/${MODULE_FILE_DIR_PATH}`

File: lib/OpenCogFunctions.cmake [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                COMMAND ${CMAKE_COMMAND} -E create_symlink "${CMAKE_CURRENT_SOURCE_DIR}/${MODULE_FILE}" "${GUILE_SYMLINK_DIR}/${MODULE_FILE_DIR_PATH}/${MODULE_FILE}"
```
Comment (linas) [inline lib/OpenCogFunctions.cmake]: You probably don't need the `${MODULE_FILE}` at the very very end; this is implicit.

File: lib/OpenCogFunctions.cmake [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            )
+            INSTALL(FILES
+                ${MODULE_FILE}
+                DESTINATION "${GUILE_INSTALL_DIR}/${MODULE_FILE_DIR_PATH}"
+            )
+        ELSE()
```
Comment (linas) [inline lib/OpenCogFunctions.cmake]: The IF and the ELSE parts seem to be identical...!?

**Comment by linas** [comment]
I like the idea of symlinks, that solves many problems.  

I am confused by the discussion of paths and modules.   A `blah.scm`  is a module if and only if it has `(define-module blah)` inside of it, *and* can be found in the load-path at an appropriate location.  Thus, for example, `utilities.scm` is not a module; it cannot be loaded by saying `(use-modules (opencog utilities))`  and the functions if might define are NOT public, unless explicitly made public or exported.   This is especially the case for the relex2logic rules files... which you can't load by saying `(use-modules (opencog nlp relex2logic rules adverbialpp))` or something like that.

**Comment by linas** [comment]
Other than that, I'm fine with this; the symlink idea alone is worth it.

File: lib/OpenCogFunctions.cmake (hunk 1 above)
Comment (AmeBel) [inline lib/OpenCogFunctions.cmake]: yap, missed that.

File: lib/OpenCogFunctions.cmake (hunk 3 above)
Comment (AmeBel) [inline lib/OpenCogFunctions.cmake]: The IF part is for `foo.scm ` in /path/opencog/foo/foo.scm` where `(define-moudle (opencog foo)) should be present in and thus should be installed/symlinked to `/path/opencog/` directory, for imports to work,  while the ELSE part is for `/path/opencog/foo/bar.scm` that will be installed/symlinked in a similar directory path.

**Comment by AmeBel** [comment]
You are right, my deiscription wasn't complete. What I meant to say was as you described.

File: lib/OpenCogFunctions.cmake (hunk 2 above)
Comment (AmeBel) [inline lib/OpenCogFunctions.cmake]: It is needed

File: lib/OpenCogFunctions.cmake (hunk 3 above)
Comment (AmeBel) [inline lib/OpenCogFunctions.cmake]: Thus there would be no need for https://github.com/opencog/opencog/tree/master/opencog/scm, each module can be self-contained in a directory.

To deal with https://github.com/opencog/opencog/blob/master/opencog/scm/opencog/atom-types.scm though, maybe each module that define a new type should be a part of the module definition. For example nlp types should be part of `(define-module (opencog nlp))`

Comment (linas) [inline lib/OpenCogFunctions.cmake]: ah, I see, there was a difference of one S in the text: `MODULE_FILE_DIR_PATH` vs. `MODULE_FILES_DIR_PATH`   and that one S was .. almost invisible.  Could you change the names to something like MODULE_FILE_PATH and MODULE_DIR_PATH  or anything, to make it easier to see that they're different?

Comment (AmeBel) [inline lib/OpenCogFunctions.cmake]: sure

**Pull request merged**